templates/openstackclient/bin/tripleo-deploy.sh

```diff
@@ -120,11 +121,48 @@ play() {
     sed -i -e 's/OSP_DIRECTOR_OPERATOR_DEPLOY_IDENTIFIER/'${DEPLOY_IDENTIFIER}'/g' $i
   done
 
+  STACK_ACTION_ARG=""
+  if [ "${OSP_VERSION}" = "16.2" ]; then
+    # We need stack_action to be set correctly in 16.2 as the network_config and octavia roles in tripleo-ansible
+    # rely on it.
+    # As we can't use the existance of a heat stack, instead use the existance of /var/lib/tripleo-config on any
+    # overcloud host (created by the tripleo-bootstrap role) to determine if this is a "CREATE" or an "UPDATE"
+    cat <<EOF > set_stack_action_playbook.yaml
+{{` ---
+- hosts: overcloud
+  name: Determine the correct stack action for 16.2 deployments
+  gather_facts: false
+  tasks:
+    - stat:
```

Comment:
So, we check any node in the overcloud group, and if the directory exists, then we set the stack_action to UPDATE for all nodes.

I just wonder if this will impact nodes during a scale out, since the Heat StackAction for those new nodes would be CREATE. Off the top of my head, I think this would be fine since the main logic around stack_action is used to manage things that exist on the Controller nodes. Computes should be ok, but it might be something worth testing to double check.

Looking at the network_config tasks, we have logic in there that should be fine even if the stack_action is set to UPDATE. It will run os-net-config if it can't find evidence of it being executed previously:

https://github.com/openstack/tripleo-ansible/blob/stable/train/tripleo_ansible/roles/tripleo-network-config/tasks/main.yml#L113-L128

Reply:
> I just wonder if this will impact nodes during a scale out, since the Heat StackAction for those new nodes would be CREATE

AFAICT for a regular scale-out or node replacement deployment StackAction would be UPDATE:
https://github.com/openstack/tripleo-common/blob/stable/train/tripleo_common/actions/deployment.py#L163

The logic here should give the same result under normal circumstances.

I think the case where the logic could differ is if a scale-out deploy limits the playbook to just the new nodes. However doing this is not strictly valid as it would not update /etc/hosts or ssh_known_hosts on the other compute nodes which is required for VM migration.

> Looking at the network_config tasks, we have logic in there that should be fine even if the stack_action is set to UPDATE.
Yes, I noticed that when looking for usage of stack_action. I did consider always setting stack_action to UPDATE as this would work for the network_config tasks. If the Octavia change could be backported I think it would work (the other related change for pacemaker restart has already been backported to train). However I think the approach here is more flexible as any existing deployments will not have pacemaker restart change.

Reply:
Ah yeah, you're right. 

In that case, I think this should be fine then. And only required for 16.2 which you're already checking for. Hopefully 17 makes life easier for everyone when RHOSP catches up with you osp-director-operator to use ephemeral heat. :)

Reply:
ack, thanks Brendan!